third_party/UTM.h

```diff
@@ -0,0 +1,256 @@
+        radian_t LongOriginRad;
+        int    ZoneNumber;
+
+        ZoneNumber = int((LongTemp + degree_t(180))/degree_t(6)) + int(1);
+
+        if( Lat >= degree_t(56.0) && Lat < degree_t(64.0) && LongTemp >= degree_t(3.0) && LongTemp < degree_t(12.0) )
+            ZoneNumber = 32;
+
+        // Special zones for Svalbard
+        if( Lat >= degree_t(72.0) && Lat < degree_t(84.0) )
+        {
+            if(      LongTemp >= degree_t(0.0)  && LongTemp <  degree_t(9.0) ) ZoneNumber = 31;
+            else if( LongTemp >= degree_t(9.0)  && LongTemp < degree_t(21.0) ) ZoneNumber = 33;
+            else if( LongTemp >= degree_t(21.0) && LongTemp < degree_t(33.0) ) ZoneNumber = 35;
+            else if( LongTemp >= degree_t(33.0) && LongTemp < degree_t(42.0) ) ZoneNumber = 37;
+        }
+        // +3 puts origin in middle of zone
+        LongOrigin = degree_t((ZoneNumber - 1)*6 - 180 + 3);
+        LongOriginRad = LongOrigin;
+
+        //compute the UTM Zone from the latitude and longitude
+        sprintf(UTMZone, "%d%c", ZoneNumber, UTMLetterDesignator(Lat));
+
+        eccPrimeSquared = (eccSquared)/(1-eccSquared);
+
+        using namespace units::math;
+
+        N = a/sqrt(1-eccSquared*sin(LatRad)*sin(LatRad));
+        T = (double)(tan(LatRad)*tan(LatRad));
+        C = eccPrimeSquared*cos(LatRad)*cos(LatRad);
+        A = cos(LatRad)*(LongRad-LongOriginRad).value();
+
+        M = a*((1 - eccSquared/4 - 3*eccSquared*eccSquared/64
+                - 5*eccSquared*eccSquared*eccSquared/256)
+                 * (double)LatRad
+               - (3*eccSquared/8 + 3*eccSquared*eccSquared/32
+                  + 45*eccSquared*eccSquared*eccSquared/1024)
+                  *sin(2*LatRad)
+               + (15*eccSquared*eccSquared/256
+                  + 45*eccSquared*eccSquared*eccSquared/1024)*sin(4*LatRad)
+               - (35*eccSquared*eccSquared*eccSquared/3072)*sin(6*LatRad));
+
+        UTMEasting = 
+        (k0*N*(A+(1-T+C)*A*A*A/6
+               + (5-18*T+T*T+72*C-58*eccPrimeSquared)*A*A*A*A*A/120)
+         + (meter_t)500000.0);
+
+        UTMNorthing = 
+        (k0*(M+N*tan(LatRad)
+             *(A*A/2+(5-T+9*C+4*C*C)*A*A*A*A/24
+               + (61-58*T+T*T+600*C-330*eccPrimeSquared)*A*A*A*A*A*A/720)));
+
+        if(Lat < degree_t(0))
+        {
+            //10000000 meter offset for southern hemisphere
+            UTMNorthing += meter_t(10000000.0);
+        }
+    }
+
+    /**
+     * Converts UTM coords to lat/long.  Equations from USGS Bulletin 1532
+     *
+     * East Longitudes are positive, West longitudes are negative.
+     * North latitudes are positive, South latitudes are negative
+     * Lat and Long are in fractional degrees.
+     *
+     * Written by Chuck Gantz- chuck.gantz@globalstar.com
+     */
+    template<typename T>
+    static inline void UTMtoLL(const T UTMNorthing, const T UTMEasting,
+                               const char* UTMZone, double& Lat,  double& Long )
```

Comment:
I've only just noticed this... This function doesn't actually need to be templated. It looks like it'll only work with ``double``s anyhow. You could just make ``UTMNorthing`` and ``UTMEasting`` ``double``s or ``meter_t``s instead. Up to you which you go for.

Reply:
Something else I forgot to mention before... I'd also add member functions for doing the conversions to ``LatLon`` and ``UTMCoordinate`` just to make it a cleaner interface. Then you could just write e.g. ``auto utm = myGPSCoords.toUTM();``.

Reply:
I am afraid I don't understand this change.

the functions 
latLonToCartesian(...) and
cartesianToLatLon(...) 

are elements of the MapCoordinate namespace and not members of any class, right?
Wouldn't it be more intuitive to have the conversion functions cartesianToUTM(...) etc. as elements of the namespace as well, instead of member functions of coordinate types?

Reply:
Yeah, kinda agree with @FullAPVayne here. For better or worse, my original design was to have free funtions for doing the conversion and light structs to hold the coordinates so I see no particular reason not to keep doing that...

Reply:
But also, this function is in a 3rd party library and, while I agree it's not an ideal signature, we probably shouldn't mess with it! My general assumption would be that things in third part remain unmodified unless we actually intend to upstream our modifications and this guy is probably not going to accept a PR adding a units requirement to his library!

Reply:
Ok, I'll change back the UTM stuff and adapt the calls in map_coordinate.h
Guess I'll be done in a bit, I'll just push again and write another comment.

Reply:
Apologies for the late reply. This all sounds sensible to me.